pymedphys/_dicom/anonymise.py

```diff
@@ -662,5 +660,5 @@ def get_anonymous_replacement_value(
     if replacement_strategy is not None:
         replacement_value = replacement_strategy[vr](current_value)
     else:
-        replacement_value = anon.get_vr_anonymous_replacement_value_dict()[vr]
+        replacement_value = anon.ANONYMISATION_HARDCODE_DISPATCH[vr](current_value)
     return replacement_value
```

Comment:
To remove the duplicated logic of `replacement_strategy[vr](current_value)` and `anon.ANONYMISATION_HARDCODE_DISPATCH[vr](current_value)` (look-up dict with `vr` and then call the function with `current_value`), can we change to the following:

```python
if replacement_strategy is not None:
    replacement_strategy = anon.ANONYMISATION_HARDCODE_DISPATCH


replacement_value = replacement_strategy[vr](current_value)
```

Reply:
I get that in Python, “Object references are passed by value” (reassignment of the argument within the function doesn't affect the parameter that was passed in, while mutation of the contents of the argument within the function does affect the contents of the parameter passed in).  But this is incredibly similar in approach to having it as a default parameter.
The default dispatch dict is in theory mutable, but in practice it is not (or at least not intended to be mutated).
Mutation of anon.ANONYMISATION_HARDCODE_DISPATCH will effect subsequent runs, regardless of how it got mutated.  If we want to be 100% sure that it does not get mutated, things get complicated (there are frozen* and there is discussion of putting an immutable dict in Python 3.9).

pushed change (it's the right thing to do, given the current limitations in python not readily supporting immutable dictionaries)

Reply:
> But this is incredibly similar in approach to having it as a default parameter.

On further thought, yeah, I don't think we've actually changed the problem. Because `anon.ANONYMISATION_HARDCODE_DISPATCH` is defined outside the function, if someone were to come along and mutate it, we would still have a bad time.

I thought the code I provided was actually logically different, but I do think at the end of the day it actually has all the same traps. The only time the pattern I described works is if the dictionary is initialised within the local scope of the function. Because we're pulling in a global constant from outside the scope of the function, we're actually falling into the very same trap.

The only thing saving us, is we're going to be responsible adults and not mutate that dictionary.

Reply:
And there's me. I learnt something new :).

Reply:
I think that by wrapping the dictionary generation with an lru_cache'd function with no parameters, it results in something that resembles a factory that produces the same dictionary always, so something like an immutable dictionary as long as people always use the function call to get to it.  I'd need to create a test to confirm that mutating the produced dictionary would in fact not affect the output of the lru_cache'd function.  An alternative is to *not* use lru_cache, but then the output dict would always be a new copy with a new reference.

we probably want to put up a sign that says "do not mutate" (or fold, or spindle) this dictionary.

Reply:
> I'd need to create a test to confirm that mutating the produced dictionary would in fact not affect the output of the lru_cache'd function.

Under the hood lru_cache creates a cache dictionary where it stores inputs as keys and outputs as values. If you mutate one of the outputs of an lru_cache it will mutate it everywhere.

In this case, I don't believe we need to take this any further. We have signaled that it is a constant with `UPPER_CASE`, I think the "We are all consenting adults" rules apply here:

https://python-guide-chinese.readthedocs.io/zh_CN/latest/writing/style.html#we-are-all-consenting-adults